File: kinotic-grind/src/main/java/org/kinotic/grind/api/model/events/TaskCompletedEvent.java
package org.kinotic.grind.api.model.events;

import org.kinotic.grind.api.model.Store;
import org.kinotic.grind.api.model.StoreType;
import tools.jackson.databind.JsonNode;

/**
 * A task finished successfully, describing what it stored in the job scope.
 *
 * @param taskPath    the task's position in the run's task tree
 * @param storeType   how the task stored its result, {@link StoreType#NONE} if it stored nothing
 * @param storedName  the name the result was stored under, or null if nothing was stored
 * @param storedValue the stored value, or null if nothing was stored. Carried for in-process
 *                    subscribers; remote watchers receive null
 * @param wireValue   the stored value serialized as JSON, or null unless the task's
 *                    {@link Store} declared wire publication. Safe to cross a serialization
 *                    boundary, so remote watchers receive it
 */
public record TaskCompletedEvent(String taskPath,
                                 StoreType storeType,
                                 String storedName,
                                 Object storedValue,
                                 JsonNode wireValue) implements JobRunEvent {
}

File: kinotic-js/workspace/packages/management-api/src/api/model/grind/events/TaskCompletedEvent.ts
import { JobRunEventType } from '@/api/model/grind/events/JobRunEventType'
import { StoreType } from '@/api/model/grind/StoreType'

/**
 * A task finished successfully, describing what it stored in the job scope. The stored object
 * itself stays in the executing process; only a task whose store declared wire publication
 * carries its value here.
 */
export class TaskCompletedEvent {

    public readonly type: JobRunEventType.TASK_COMPLETED = JobRunEventType.TASK_COMPLETED

    /**
     * The task's position in the run's task tree.
     */
    public readonly taskPath: string = ''

    /**
     * How the task stored its result, NONE if it stored nothing.
     */
    public readonly storeType: StoreType = StoreType.NONE

    /**
     * The name the result was stored under, or null if nothing was stored.
     */
    public readonly storedName: string | null = null

    /**
     * The stored value as JSON, or null unless the task's store declared wire publication.
     */
    public readonly wireValue: any = null

}
